components/raftstore/src/store/snapshot_backup.rs

```diff
@@ -0,0 +1,345 @@
+use tikv_util::{info, warn};
+use tokio::sync::oneshot;
+
+use super::{metrics, PeerMsg, RaftRouter, SignificantMsg, SignificantRouter};
+use crate::coprocessor::{
+    AdminObserver, BoxAdminObserver, BoxQueryObserver, Coprocessor, CoprocessorHost,
+    Error as CopError, QueryObserver,
+};
+
+fn epoch_second_coarse() -> u64 {
+    let spec = tikv_util::time::monotonic_coarse_now();
+    spec.sec as u64
+}
+
+#[derive(Debug, Clone)]
+pub struct SnapshotBrWaitApplyRequest {
+    pub syncer: SnapshotBrWaitApplySyncer,
+    pub expected_epoch: Option<RegionEpoch>,
+    pub abort_when_term_change: bool,
+}
+
+impl SnapshotBrWaitApplyRequest {
+    /// Create a "relax" request for waiting apply.
+    /// This only waits to the last index, without checking the region epoch or
+    /// leadership migrating.
+    pub fn relaxed(syncer: SnapshotBrWaitApplySyncer) -> Self {
+        Self {
+            syncer,
+            expected_epoch: None,
+            abort_when_term_change: false,
+        }
+    }
+
+    /// Create a "strict" request for waiting apply.
+    /// This will wait to last applied index, and aborts if the region epoch not
+    /// match or the last index may not be committed.
+    pub fn strict(syncer: SnapshotBrWaitApplySyncer, epoch: RegionEpoch) -> Self {
+        Self {
+            syncer,
+            expected_epoch: Some(epoch),
+            abort_when_term_change: true,
+        }
+    }
+}
+
+pub trait SnapshotBrHandle: Sync + Send + Clone {
+    fn send_wait_apply(&self, region: u64, req: SnapshotBrWaitApplyRequest) -> crate::Result<()>;
+    fn broadcast_wait_apply(&self, req: SnapshotBrWaitApplyRequest) -> crate::Result<()>;
+    fn broadcast_check_pending_admin(
+        &self,
+        tx: UnboundedSender<CheckAdminResponse>,
+    ) -> crate::Result<()>;
+}
+
+impl<EK: KvEngine, ER: RaftEngine> SnapshotBrHandle for Arc<Mutex<RaftRouter<EK, ER>>> {
+    fn send_wait_apply(&self, region: u64, req: SnapshotBrWaitApplyRequest) -> crate::Result<()> {
+        let msg = SignificantMsg::SnapshotBrWaitApply(req);
+        metrics::SNAP_BR_WAIT_APPLY_EVENT.sent.inc();
+        self.lock().unwrap().significant_send(region, msg)
+    }
+
+    fn broadcast_wait_apply(&self, req: SnapshotBrWaitApplyRequest) -> crate::Result<()> {
+        let msg_gen = || {
+            metrics::SNAP_BR_WAIT_APPLY_EVENT.sent.inc();
+            PeerMsg::SignificantMsg(SignificantMsg::SnapshotBrWaitApply(req.clone()))
+        };
+        self.lock().unwrap().broadcast_normal(msg_gen);
+        Ok(())
+    }
+
+    fn broadcast_check_pending_admin(
+        &self,
+        tx: UnboundedSender<CheckAdminResponse>,
+    ) -> crate::Result<()> {
+        self.lock().unwrap().broadcast_normal(|| {
+            PeerMsg::SignificantMsg(SignificantMsg::CheckPendingAdmin(tx.clone()))
+        });
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+pub struct RejectIngestAndAdmin {
+    before: AtomicU64,
```

Comment:
give it a meaningful name

Reply:
Do you have some good ideas? Perhaps `PrepareDiskSnapObserver`?